src/core/server/server.cc

```diff
@@ -86,6 +88,224 @@
+    if (listener_->started_) return;
+    listener_->started_ = true;
+  }
+  listener_->StartImpl();
+}
+
+void Server::ListenerInterface::ConfigFetcherWatcher::StopServing() {
+  absl::flat_hash_set<OrphanablePtr<ListenerInterface::LogicalConnection>>
+      connections;
+  {
+    MutexLock lock(&listener_->mu_);
+    listener_->is_serving_ = false;
+    connections = std::move(listener_->connections_);
+  }
+  // Send GOAWAYs on the transports so that they disconnect when existing
+  // RPCs finish.
+  for (auto& connection : connections) {
+    connection->SendGoAway();
+  }
+}
+
+//
+// Server::ListenerInterface::LogicalConnection
+//
+
+void Server::ListenerInterface::LogicalConnection::SendGoAway() {
+  {
+    if (!SendGoAwayImpl()) {
+      return;
+    }
+    MutexLock lock(&mu_);
+    if (drain_grace_timer_handle_cancelled_) {
+      return;
+    }
+    CHECK(drain_grace_timer_handle_ == EventEngine::TaskHandle::kInvalid);
+    drain_grace_timer_handle_ = event_engine_->RunAfter(
+        std::max(Duration::Zero(),
+                 listener_->server_->channel_args()
+                     .GetDurationFromIntMillis(
+                         GRPC_ARG_SERVER_CONFIG_CHANGE_DRAIN_GRACE_TIME_MS)
+                     .value_or(Duration::Minutes(10))),
+        [self = Ref(DEBUG_LOCATION, "drain_grace_timer")]() mutable {
+          ApplicationCallbackExecCtx callback_exec_ctx;
+          ExecCtx exec_ctx;
+          // If the drain_grace_timer_ was not cancelled, disconnect
+          // immediately.
+          bool disconnect_immediately = false;
+          {
+            MutexLock lock(&self->mu_);
+            self->drain_grace_timer_handle_ = EventEngine::TaskHandle::kInvalid;
+            if (!self->drain_grace_timer_handle_cancelled_) {
+              disconnect_immediately = true;
+            }
+          }
+          if (disconnect_immediately) {
+            self->DisconnectImmediatelyImpl();
+          }
+          self.reset(DEBUG_LOCATION, "drain_grace_timer");
+        });
+  }
+}
+
+void Server::ListenerInterface::LogicalConnection::CancelDrainGraceTimer() {
+  MutexLock lock(&mu_);
+  drain_grace_timer_handle_cancelled_ = true;
+  if (drain_grace_timer_handle_ != EventEngine::TaskHandle::kInvalid) {
+    event_engine_->Cancel(drain_grace_timer_handle_);
+    drain_grace_timer_handle_ = EventEngine::TaskHandle::kInvalid;
+  }
+}
+
+//
+// Server::ListenerInterface
+//
+
+void Server::ListenerInterface::Orphan() {
+  // Cancel the watch before shutting down so as to avoid holding a ref to the
+  // listener in the watcher.
+  if (config_fetcher_watcher_ != nullptr) {
+    CHECK_NE(server_->config_fetcher(), nullptr);
+    server_->config_fetcher()->CancelWatch(config_fetcher_watcher_);
+  }
+  absl::flat_hash_set<OrphanablePtr<LogicalConnection>> connections;
+  {
+    MutexLock lock(&mu_);
+    // Orphan the connections so that they can start cleaning up.
+    connections = std::move(connections_);
+    is_serving_ = false;
```

Comment:
The old code was simply broken. The cv it waited on was for the tcp server to start, but the condition on the loop was `is_serving_ && !started_` but just before that we set `is_serving_` to false :D

Anyway, I'm fixing this by simply grabbing the ref to the tcp server while we are starting so that we don't run into issues.